src/algebra/ordered_monoid_lemmas.lean

```diff
@@ -724,3 +733,63 @@ lemma strict_mono.mul_monotone' (hf : strict_mono f) (hg : monotone g) :
 λ x y h, mul_lt_mul_of_lt_of_le (hf h) (hg h.le)
 
 end strict_mono
+
+/--
+An element `a : α` is `mul_le_cancellable` if `x ↦ a * x` is order-reflecting.
+We will make a separate version of many lemmas that require `[contravariant_class α α (*) (≤)]` with
+`mul_le_cancellable` assumptions instead. These lemmas can then be instantiated to specific types,
+like `ennreal`, where we can replace the assumption `add_le_cancellable x` by `x ≠ ∞`.
```

Comment:
I like this!  In fact, you can probably also use this (and the similarly specialized `covariant`) for monotonicity of multiplying by the positive elements.  This was not the way I was going to go with the order refactor: I had in mind of defining the "correct" subtype for which all elements had monotone multiplication, but your version seems to be more flexible.  I think that this might be the way to go for the order refactor.

One comment about the doc-string: why "order-reflecting"?  I would say that the order is preserved, since the positions of `b c` before and after are the same with respect to the inequality.  Isn't this a "relative" covariant, where you only assume monotonicity for a sigle element and not for the whole type?

Reply:
Hmm... I hadn't thought about using this for lemmas about rings/fields, but yeah, this could be useful. They probably still need to be specialized to the case where `add_le_cancellable` is replaced by `(≠ 0)`.

Order-reflecting is a term from order theory. Order-preserving/monotonicity would be the reverse implication. See
https://en.wikipedia.org/wiki/Order_theory#Functions_between_orders